packages/vite/src/node/plugins/workerImportMetaUrl.ts

```diff
@@ -0,0 +1,96 @@
+import { getAssetHash, fileToUrl } from './asset'
+import {
+  cleanUrl,
+  injectQuery,
+  multilineCommentsRE,
+  singlelineCommentsRE
+} from '../utils'
+import path from 'path'
+import { parseWorkerRequest, bundleWorkerEntry } from './worker'
+import { ENV_PUBLIC_PATH } from '../constants'
+import MagicString from 'magic-string'
+
+const WorkerFileId = 'worker_url_file'
+
+export function workerImportMetaUrlPlugin(config: ResolvedConfig): Plugin {
+  const isBuild = config.command === 'build'
+
+  return {
+    name: 'vite:worker-import-meta-url',
+
+    async transform(code, id, options) {
+      const query = parseWorkerRequest(id)
+      if (query && query[WorkerFileId] != null) {
+        return {
+          code: `import '${ENV_PUBLIC_PATH}'\n` + code
+        }
+      }
+      if (
+        (code.includes('new Worker') || code.includes('new ShareWorker')) &&
+        code.includes('new URL') &&
+        code.includes(`import.meta.url`)
+      ) {
+        const importMetaUrlRE =
+          /\bnew\s+(Worker|SharedWorker)+\s*\(\s*(new\s+URL\s*\(\s*('[^']+'|"[^"]+"|`[^`]+`)\s*,\s*import\.meta\.url\s*\))/g
+        const noCommentsCode = code
+          .replace(multilineCommentsRE, (m) => ' '.repeat(m.length))
+          .replace(singlelineCommentsRE, (m) => ' '.repeat(m.length))
+        let match: RegExpExecArray | null
+        let s: MagicString | null = null
+        while ((match = importMetaUrlRE.exec(noCommentsCode))) {
+          const { 0: allExp, 2: exp, 3: rawUrl, index } = match
+
+          const urlIndex = allExp.indexOf(exp) + index
+
+          if (options?.ssr) {
+            this.error(
+              `\`new URL(url, import.meta.url)\` is not supported in SSR.`,
+              urlIndex
+            )
+          }
+
+          // potential dynamic template string
+          if (rawUrl[0] === '`' && /\$\{/.test(rawUrl)) {
+            this.error(
+              `\`new URL(url, import.meta.url)\` is not supported in dynamic template string.`,
+              urlIndex
+            )
+          }
+
+          if (!s) s = new MagicString(code)
+
+          const basename = rawUrl.slice(1, -1)
+          const file = path.resolve(path.dirname(id), basename)
+          let url: string
+          if (isBuild) {
+            const content = await bundleWorkerEntry(config, file)
+            const basename = path.parse(cleanUrl(file)).name
+            const contentHash = getAssetHash(content)
+            const fileName = path.posix.join(
+              config.build.assetsDir,
+              `${basename}.${contentHash}.js`
+            )
+            url = `__VITE_ASSET__${this.emitFile({
+              fileName,
+              type: 'asset',
+              source: content
+            })}__`
+          } else {
+            url = await fileToUrl(cleanUrl(file), config, this)
+            url = injectQuery(url, WorkerFileId)
+          }
+          s.overwrite(urlIndex, urlIndex + exp.length, JSON.stringify(url))
```

Comment:
```suggestion
          s.overwrite(
            index,
            index + exp.length,
            `new ${workerConstructor}(new URL(${JSON.stringify(
              url
            )}, self.location)`
          )
```

1. replace from `index` to avoid find `urlIndex`
2. we need remain the `self.location`, according the note from here https://github.com/vitejs/vite/blob/aab303f7bd333307c77363259f97a310762c4848/packages/vite/src/node/plugins/assetImportMetaUrl.ts#L48-L51